File: android/src/main/java/com/navigine/naviginesdk/IconMapObjectController.java
package com.navigine.naviginesdk;

import android.content.Context;

import com.navigine.idl.java.IconMapObject;
import com.navigine.idl.java.LocationWindow;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;

public class IconMapObjectController {
  private final BinaryMessenger binaryMessenger;
  private final Context context;
  private final LocationWindow locationWindow;
  private final Map<Integer, NavigineIconMapObject> mapObjects = new HashMap<>();

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public IconMapObjectController(
          BinaryMessenger messenger,
          Context context,
          LocationWindow locationWindow
  ) {
    this.binaryMessenger = messenger;
    this.context = context;
    this.locationWindow = locationWindow;
  }

  public int addIconMapObject() {
    IconMapObject iconMapObject = locationWindow.addIconMapObject();
    int iconMapObjectId = iconMapObject.getId();
    NavigineIconMapObject navigineIconMapObject = new NavigineIconMapObject(
            iconMapObjectId,
            this.context,
            iconMapObject,
            binaryMessenger);

    mapObjects.put(iconMapObjectId, navigineIconMapObject);
    return iconMapObjectId;
  }

  public boolean removeIconMapObject(int id) {
    NavigineIconMapObject iconMapObject = mapObjects.get(id);
    if (iconMapObject == null) {
      return false;
    }

    boolean success = locationWindow.removeIconMapObject(iconMapObject.getMapObject());
    mapObjects.remove(id);
    return success;
  }


}

File: ios/Classes/IconMapObjectController.swift
import Navigine

class IconMapObjectController: NSObject {
    public weak var locationView: NCLocationView?

    private let pluginRegistrar: FlutterPluginRegistrar!
    private var mapObjects: [Int32: NavigineIconMapObject] = [:]

    public required init(
        registrar: FlutterPluginRegistrar,
        locationView: NCLocationView
    ) {
        self.pluginRegistrar = registrar
        self.locationView = locationView
        super.init()
    }

    public func addIconMapObject() -> Int32? {
        if locationView == nil {
            return nil
        }
        let iconMapObject = locationView!.addIconMapObject()
        let iconMapObjectId = iconMapObject.getId()
        mapObjects[iconMapObjectId] = NavigineIconMapObject(id: iconMapObjectId, mapObject: iconMapObject, registrar: self.pluginRegistrar)
        return iconMapObjectId

    }
    
    public func removeIconMapObject(id: Int32) -> Bool? {
        if locationView == nil {
            return nil
        }
        let iconMapObject = mapObjects[id]
        if (iconMapObject == nil) {
            return false
        }
        
        let success = locationView?.remove((iconMapObject?.getMapObject())!);
        mapObjects.removeValue(forKey: id)
        return success
    }
}
